app/src/main/java/org/schabi/newpipe/ReCaptchaActivity.java

```diff
@@ -89,74 +91,89 @@ protected void onCreate(Bundle savedInstanceState) {
+    public boolean onCreateOptionsMenu(Menu menu) {
+        boolean ret = super.onCreateOptionsMenu(menu);
 
-        ReCaptchaWebViewClient(Activity ctx) {
-            context = ctx;
+        ActionBar actionBar = getSupportActionBar();
+        if (actionBar != null) {
+            actionBar.setDisplayHomeAsUpEnabled(true);
+            actionBar.setHomeAsUpIndicator(getResources().getDrawable(R.drawable.ic_done_white_24dp));
+            actionBar.setTitle(R.string.title_activity_recaptcha);
+            actionBar.setSubtitle(R.string.subtitle_activity_recaptcha);
         }
 
-        @Override
-        public void onPageStarted(WebView view, String url, Bitmap favicon) {
-            // TODO: Start Loader
-            super.onPageStarted(view, url, favicon);
+        return ret;
+    }
+
+    @Override
+    public void onBackPressed() {
+        saveCookiesAndFinish();
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        int id = item.getItemId();
+        switch (id) {
+            case android.R.id.home:
+                saveCookiesAndFinish();
+                return true;
+            default:
+                return false;
         }
+    }
 
-        @Override
-        public void onPageFinished(WebView view, String url) {
-            String cookies = CookieManager.getInstance().getCookie(url);
+    private void saveCookiesAndFinish() {
+        if (!foundCookies.isEmpty()) {
+            // Give cookies to Downloader class
+            DownloaderImpl.getInstance().setCookies(foundCookies);
+            setResult(RESULT_OK);
+        }
 
-            // TODO: Stop Loader
+        Intent intent = new Intent(this, org.schabi.newpipe.MainActivity.class);
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+        NavUtils.navigateUpTo(this, intent);
+    }
 
-            // find cookies : s_gl & goojf and Add cookies to Downloader
-            if (find_access_cookies(cookies)) {
-                // Give cookies to Downloader class
-                DownloaderImpl.getInstance().setCookies(mCookies);
 
-                // Closing activity and return to parent
-                setResult(RESULT_OK);
-                finish();
-            }
-        }
 
-        private boolean find_access_cookies(String cookies) {
-            boolean ret = false;
-            String c_s_gl = "";
-            String c_goojf = "";
-
-            String[] parts = cookies.split("; ");
-            for (String part : parts) {
-                if (part.trim().startsWith("s_gl")) {
-                    c_s_gl = part.trim();
-                }
-                if (part.trim().startsWith("goojf")) {
-                    c_goojf = part.trim();
-                }
+    private void handleCookies(@Nullable String cookies) {
+        if (MainActivity.DEBUG) Log.d(TAG, "handleCookies: cookies=" + (cookies == null ? "null" : cookies));
+        if (cookies == null) return;
+
+        addYoutubeCookies(cookies);
```

Comment:
It might be good to check the service id first. Just in case we get a reCAPTCHA from a different service.

Reply:
This could be done, but it would require modifying the downloader making it service-aware, so that it can throw service-aware recaptcha exceptions. Another option would be to recognize the service from the url using its url handler, but API urls wouldn't be probably recognized.
So I think this is good as it is, since if the `addYoutubeCookies()` function is able to find YouTube cookies they get saved, otherwise they do not. And, as the comment on the line below suggests, it is simple to add new cookie extractors for other services in the future.

Reply:
Okay :)